src/neo/SmartContract/Native/ContractManagement.cs

```diff
@@ -104,16 +105,21 @@ internal override async ContractTask OnPersist(ApplicationEngine engine)
             foreach (NativeContract contract in Contracts)
             {
                 uint[] updates = engine.ProtocolSettings.NativeUpdateHistory[contract.Name];
```

Comment:
I have another solution using UpdateCounter, but it require to deserialize all the native contracts in each block

```csharp
uint[] updates = engine.ProtocolSettings.NativeUpdateHistory[contract.Name];
                StorageKey contractKey = CreateStorageKey(Prefix_Contract).Add(contract.Hash);
                ContractState storedContract = engine.Snapshot.TryGet(contractKey)?.GetInteroperable<ContractState>();
                ushort version = storedContract?.UpdateCounter ?? 0;
                if (updates.Length >= version) continue;
                if (updates[version] != engine.PersistingBlock.Index)
                    continue;
                engine.Snapshot.Add(contractKey, new StorageItem(new ContractState
                {
                    Id = contract.Id,
                    Nef = contract.Nef,
                    Hash = contract.Hash,
                    UpdateCounter = (ushort)(version + 1),
                    Manifest = contract.Manifest.ForVersion(version) // Get versioned Abi
                }));
                await contract.Initialize(engine);
                engine.SendNotification(Hash, version != 0 ? "Update" : "Deploy", new VM.Types.Array { contract.Hash.ToArray(), version });
```